### PR TITLE
Improve Merge performance

### PR DESCRIPTION
**Pull request by brian-brazil**
Use a heap for Next for merges, and
pre-compute if there's many postings on the
unset path.

Add posting lookup benchmarks

Signed-off-by: Brian Brazil <brian.brazil@robustperception.io>

Fixes https://github.com/prometheus/prometheus/issues/5173

FYI @jacksontj @bwplotka

### CODE REVIEW

**Comment by brian-brazil** [comment]
Benchmarks relative to before #486. The extremely slow case is still fixed, and some fairly fast cases get a little slower. I do have ideas for further optimisations.

```
benchmark                                                            old ns/op        new ns/op      delta
BenchmarkHeadPostingForMatchers/n="1"-4                              5420             5742           +5.94%
BenchmarkHeadPostingForMatchers/n="1",j="foo"-4                      6994             6531           -6.62%
BenchmarkHeadPostingForMatchers/j="foo",n="1"-4                      6038             5512           -8.71%
BenchmarkHeadPostingForMatchers/n="1",j!="foo"-4                     10364            10272          -0.89%
BenchmarkHeadPostingForMatchers/i=~".*"-4                            3292078491       3565717098     +8.31%
BenchmarkHeadPostingForMatchers/i=~".+"-4                            4104045278       5388378129     +31.29%
BenchmarkHeadPostingForMatchers/i=~""-4                              172650291999     1806584165     -98.95%
BenchmarkHeadPostingForMatchers/i!=""-4                              3715780926       4580742079     +23.28%
BenchmarkHeadPostingForMatchers/n="1",i=~".*",j="foo"-4              112866419        152246610      +34.89%
BenchmarkHeadPostingForMatchers/n="1",i=~".*",i!="2",j="foo"-4       196685677        214845649      +9.23%
BenchmarkHeadPostingForMatchers/n="1",i!="",j="foo"-4                264144048        268635277      +1.70%
BenchmarkHeadPostingForMatchers/n="1",i=~".+",j="foo"-4              292865856        292393747      -0.16%
BenchmarkHeadPostingForMatchers/n="1",i=~"1.+",j="foo"-4             113954164        122264453      +7.29%
BenchmarkHeadPostingForMatchers/n="1",i=~".+",i!="2",j="foo"-4       383920764        357080814      -6.99%
BenchmarkHeadPostingForMatchers/n="1",i=~".+",i!~"2.*",j="foo"-4     411214259        509801731      +23.97%

benchmark                                                            old allocs     new allocs     delta
BenchmarkHeadPostingForMatchers/n="1"-4                              8              8              +0.00%
BenchmarkHeadPostingForMatchers/n="1",j="foo"-4                      11             11             +0.00%
BenchmarkHeadPostingForMatchers/j="foo",n="1"-4                      11             11             +0.00%
BenchmarkHeadPostingForMatchers/n="1",j!="foo"-4                     19             19             +0.00%
BenchmarkHeadPostingForMatchers/i=~".*"-4                            62             68             +9.68%
BenchmarkHeadPostingForMatchers/i=~".+"-4                            200119         100123         -49.97%
BenchmarkHeadPostingForMatchers/i=~""-4                              200996         100126         -50.19%
BenchmarkHeadPostingForMatchers/i!=""-4                              200111         100114         -49.97%
BenchmarkHeadPostingForMatchers/n="1",i=~".*",j="foo"-4              20             20             +0.00%
BenchmarkHeadPostingForMatchers/n="1",i=~".*",i!="2",j="foo"-4       30             31             +3.33%
BenchmarkHeadPostingForMatchers/n="1",i!="",j="foo"-4                200076         100110         -49.96%
BenchmarkHeadPostingForMatchers/n="1",i=~".+",j="foo"-4              200076         100110         -49.96%
BenchmarkHeadPostingForMatchers/n="1",i=~"1.+",j="foo"-4             22274          11176          -49.82%
BenchmarkHeadPostingForMatchers/n="1",i=~".+",i!="2",j="foo"-4       200086         100109         -49.97%
BenchmarkHeadPostingForMatchers/n="1",i=~".+",i!~"2.*",j="foo"-4     222345         111293         -49.95%

benchmark                                                            old bytes     new bytes     delta
BenchmarkHeadPostingForMatchers/n="1"-4                              1360          1360          +0.00%
BenchmarkHeadPostingForMatchers/n="1",j="foo"-4                      1392          1392          +0.00%
BenchmarkHeadPostingForMatchers/j="foo",n="1"-4                      1392          1392          +0.00%
BenchmarkHeadPostingForMatchers/n="1",j!="foo"-4                     1616          1616          +0.00%
BenchmarkHeadPostingForMatchers/i=~".*"-4                            106202232     106284872     +0.08%
BenchmarkHeadPostingForMatchers/i=~".+"-4                            132704968     129518904     -2.40%
BenchmarkHeadPostingForMatchers/i=~""-4                              33211096      113464616     +241.65%
BenchmarkHeadPostingForMatchers/i!=""-4                              132696208     129501952     -2.41%
BenchmarkHeadPostingForMatchers/n="1",i=~".*",j="foo"-4              1608950       1609769       +0.05%
BenchmarkHeadPostingForMatchers/n="1",i=~".*",i!="2",j="foo"-4       3216492       3214998       -0.05%
BenchmarkHeadPostingForMatchers/n="1",i!="",j="foo"-4                28106771      73079923      +160.01%
BenchmarkHeadPostingForMatchers/n="1",i=~".+",j="foo"-4              28108409      73078400      +159.99%
BenchmarkHeadPostingForMatchers/n="1",i=~"1.+",j="foo"-4             4149430       5434900       +30.98%
BenchmarkHeadPostingForMatchers/n="1",i=~".+",i!="2",j="foo"-4       29713221      58628874      +97.32%
BenchmarkHeadPostingForMatchers/n="1",i=~".+",i!~"2.*",j="foo"-4     32256656      70100792      +117.32%
```

**Comment by tomwilkie** [comment]
We should talk: https://github.com/cortexproject/cortex/blob/master/pkg/querier/batch/merge.go#L11

;-)

File: index/postings.go [hunk 1]
```diff
@@ -365,25 +366,130 @@ func Merge(its ...Postings) Postings {
 	if len(its) == 1 {
 		return its[0]
 	}
-	// All the uses of this function immediately expand it, so
-	// collect everything in a map. This is more efficient
-	// when there's 100ks of postings, compared to
-	// having a tree of merge objects.
-	pm := make(map[uint64]struct{}, len(its))
-	for _, it := range its {
-		for it.Next() {
```
[the remark this answers is not in the document]
Comment (brian-brazil) [reply index/postings.go]: It's not the most efficient way to do things, as it's not taking advantage of the runs within each iterator

File: querier.go [hunk 2]
```diff
@@ -354,11 +354,22 @@ func postingsForUnsetLabelMatcher(ix IndexReader, m labels.Matcher) (index.Posti
 		rit = append(rit, it)
 	}
 
+	merged := index.Merge(rit...)
+	// With many many postings, it's best to pre-calculate
+	// the merged list via next rather than have a ton of seeks.
```
[the remark this answers is not in the document]
Comment (brian-brazil) [reply querier.go]: Seek is inefficient if you have to do it on hundreds of iterators

File: index/postings.go [hunk 3]
```diff
@@ -365,25 +366,130 @@ func Merge(its ...Postings) Postings {
...
+		return false
+	}
+
+	if !it.initilized {
+		it.cur = it.h[0].At()
+		it.initilized = true
+		return true
+	}
+	if !it.heaped {
+		heap.Init(&(it.h))
```
[the remark this answers is not in the document]
Comment (brian-brazil) [reply index/postings.go]: Fixed

File: index/postings.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (krasi-georgiev) [reply index/postings.go]: yep I see it now, thanks!

File: querier.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (krasi-georgiev) [reply querier.go]: My question was more on how does this avoid a ton of seeks.

[the remark this answers is not in the document]
Comment (brian-brazil) [reply querier.go]: Without and Intersect use Seeks, by changing to a pre-calculated list they happen (efficiently) on that list rather than on the Heap.

File: index/postings.go [hunk 4]
```diff
@@ -365,25 +366,130 @@ func Merge(its ...Postings) Postings {
...
 	}
-	pl := make([]uint64, 0, len(pm))
-	for p := range pm {
-		pl = append(pl, p)
+	if it.cur >= id {
+		return true
 	}
-	sort.Slice(pl, func(i, j int) bool { return pl[i] < pl[j] })
-	return newListPostings(pl)
+	// Heapifying when there is lots of Seeks is inefficient.
```
[the remark this answers is not in the document]
Comment (krasi-georgiev) [reply index/postings.go]: btw shouldn't the heap me marked as dirty only when at least one i.Seek has succeeded?

```
if i.Seek(id) {
	if it.heaped {
		it.heaped = false
	}
```

File: index/postings.go [hunk 5]
```diff
@@ -365,25 +366,130 @@ func Merge(its ...Postings) Postings {
...
+	if !it.initilized {
+		it.cur = it.h[0].At()
+		it.initilized = true
+		return true
+	}
+
+	for {
+		cur := it.h[0]
+		if !cur.Next() {
+			heap.Pop(&(it.h))
```
[the remark this answers is not in the document]
Comment (brian-brazil) [reply index/postings.go]: That won't work, as the `&` would apply to `it`

File: index/postings.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (brian-brazil) [reply index/postings.go]: In that case, the heap is now empty so it makes no difference.

**Comment by brian-brazil** [comment]
Updated.

File: index/postings.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (krasi-georgiev) [reply index/postings.go]: that is weird when I do this it always prints the same pointer.
https://play.golang.org/p/VZVWfCBw-wm

and even when I add :  `fmt.Printf("Fix %p , %p\n", &(it.h), &it.h)` and run the `TestMergedPostings` it does indeed print the same pointer so it seems that the extra brackets don't make any difference.

File: index/postings.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (krasi-georgiev) [reply index/postings.go]: sorry I didn't understand what you mean.

File: index/postings.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (brian-brazil) [reply index/postings.go]: Hmm, it does work. Guess I was confusing it with something else.

File: index/postings.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (brian-brazil) [reply index/postings.go]: It all the Seeks return false, that means we've gotten to the end of all the iterators.

File: index/postings.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (krasi-georgiev) [reply index/postings.go]: the strange thing is that even `&it` and `&it.h` both print the same pointer so I asked in the golang channel and the answer was : "even for embedded struct - all start at the same location in memory."

so I guess there is no way to get a pointer of the actual child fields.

File: index/postings.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (krasi-georgiev) [reply index/postings.go]: yes that is why I thought that `heap.Init(&it.h)` should be run only when at least one seek has succeed. Are we talking about the same thing?

[the remark this answers is not in the document]
Comment (brian-brazil) [reply index/postings.go]: Yes, but it's harmless. Heaping a zero-length array is fast.

[the remark this answers is not in the document]
Comment (krasi-georgiev) [reply index/postings.go]: ok, would you mind adding a bit more detail in the comment like.

// Heapifying when there is lots of Seeks is inefficient so
// just mark it as dirty and to be re-heapified on the Next() call.

File: index/postings.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (krasi-georgiev) [reply index/postings.go]: after some more details from the golang channel it seems that just `heap.Pop(&it.h)` will work so all good.

[the remark this answers is not in the document]
Comment (brian-brazil) [reply index/postings.go]: I was thinking more the typing may have been an issue.

File: index/postings.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (brian-brazil) [reply index/postings.go]: Done

File: head_bench_test.go [hunk 6]
```diff
@@ -56,17 +56,56 @@ func BenchmarkHeadPostingForMatchers(b *testing.B) {
 	testutil.Ok(b, err)
 	defer h.Close()
```
Comment (krasi-georgiev) [inline head_bench_test.go]:
```suggestion
	defer func() {
		testutil.Ok(b, h.Close())
	}()
```

File: head_bench_test.go [hunk 7]
```diff
@@ -56,17 +56,56 @@ func BenchmarkHeadPostingForMatchers(b *testing.B) {
...
 
-	// TODO: vary number of series
-	for i := 0; i < 1000000; i++ {
-		h.getOrCreate(uint64(i), labels.FromStrings("a", strconv.Itoa(i)))
+	var hash uint64
+	for n := 0; n < 10; n++ {
+		for i := 0; i < 100000; i++ {
+			h.getOrCreate(hash, labels.FromStrings("i", strconv.Itoa(i), "n", strconv.Itoa(i), "j", "foo"))
+			hash++
+			h.getOrCreate(hash, labels.FromStrings("i", strconv.Itoa(i), "n", strconv.Itoa(i), "j", "bar"))
```
Comment (krasi-georgiev) [inline head_bench_test.go]: Why is this needed?

Comment (brian-brazil) [reply head_bench_test.go]: So there's some series that won't be matched.

Comment (krasi-georgiev) [reply head_bench_test.go]: yep I understand this part, but why is it needed? 
I would put a comment as well.

Comment (brian-brazil) [reply head_bench_test.go]: Done

**Comment by krasi-georgiev** [comment]
LGTM, thanks Brian!

File: querier.go [hunk 8]
```diff
@@ -354,11 +354,23 @@ func postingsForUnsetLabelMatcher(ix IndexReader, m labels.Matcher) (index.Posti
 		rit = append(rit, it)
 	}
 
+	merged := index.Merge(rit...)
+	// With many many postings, it's best to pre-calculate
```
Comment (bwplotka) [inline querier.go]: Nice, ping for me to update Thanos code as well: https://github.com/improbable-eng/thanos/blob/master/pkg/store/bucket.go#L1249
